src/routing/middleware/cors.rs

```diff
@@ -1,80 +1,544 @@
+        assert!(resp.headers().get(ACCESS_CONTROL_ALLOW_METHODS).is_none());
+        assert!(resp.headers().get(ACCESS_CONTROL_ALLOW_HEADERS).is_none());
+        assert!(resp.headers().get(ACCESS_CONTROL_MAX_AGE).is_none());
+
+        // Check that default response is sent for OPTIONS request if the backend returns incorrect response
+        let router = Router::new()
+            .fallback(|| async { StatusCode::METHOD_NOT_ALLOWED })
+            .layer(from_fn_with_state(s.clone(), middleware));
+
+        let mut req = Request::new(Body::empty());
+        *req.method_mut() = Method::OPTIONS;
+        req.extensions_mut().insert(canister_id);
+        let resp = router.oneshot(req).await.unwrap();
+        assert_eq!(resp.status(), StatusCode::OK);
+
+        // Make sure the canister is marked as invalid
+        assert!(s.invalid_canisters.contains_key(&canister_id));
+
+        assert_eq!(
+            resp.headers()
+                .get(ACCESS_CONTROL_ALLOW_HEADERS)
+                .unwrap()
+                .to_str()
+                .unwrap(),
+            s.allow_headers.to_str().unwrap()
+        );
+        assert_eq!(
+            resp.headers()
+                .get(ACCESS_CONTROL_ALLOW_METHODS)
+                .unwrap()
+                .to_str()
+                .unwrap(),
+            s.allow_methods.to_str().unwrap(),
+        );
+        assert_eq!(
+            resp.headers()
+                .get(ACCESS_CONTROL_MAX_AGE)
+                .unwrap()
+                .to_str()
+                .unwrap(),
+            s.max_age,
+        );
+        assert_eq!(
+            resp.headers()
+                .get(ACCESS_CONTROL_ALLOW_ORIGIN)
+                .unwrap()
+                .to_str()
+                .unwrap(),
+            s.allow_origin,
+        );
+        assert_eq!(resp.headers().get(VARY).unwrap().to_str().unwrap(), s.vary);
+        assert!(resp.headers().get(ACCESS_CONTROL_EXPOSE_HEADERS).is_none());
+    }
+
+    #[test]
+    fn test_is_valid_preflight_response() {
+        // Check ok
+        let mut r = Response::new(Body::empty());
+        r.headers_mut().insert(
+            ACCESS_CONTROL_ALLOW_HEADERS,
+            HeaderValue::from_static("foo"),
+        );
+        assert!(is_valid_preflight_response(&r));
+
+        // Check no headers
+        let r = Response::new(Body::empty());
+        assert!(!is_valid_preflight_response(&r));
+
+        // Check non-empty body
+        let mut r = Response::new(Body::new(http_body_util::Full::new(Bytes::from_static(
+            "foo".as_bytes(),
+        ))));
+        r.headers_mut().insert(
+            ACCESS_CONTROL_ALLOW_HEADERS,
+            HeaderValue::from_static("foo"),
+        );
+        assert!(!is_valid_preflight_response(&r));
+
+        // Check bad status
+        let mut r = Response::new(Body::empty());
+        *r.status_mut() = StatusCode::METHOD_NOT_ALLOWED;
+        assert!(!is_valid_preflight_response(&r));
```

Comment:
assert!(
        !is_valid_preflight_response(&r),
        "Expected invalid preflight response due to incorrect status code, but it was valid"
    );